File: desktop/src/api/auditoria.ts
import { invoke } from "@tauri-apps/api/core";
import type { IncidenteGafete } from "./gafetes";
import type { CargaCompleta } from "./comun";

// Espejo de comandos/auditoria.rs y
// src/database/queries/auditoria.rs (`CambioAuditado`) del núcleo. Sin
// paginado a propósito — mismo criterio que Historial: trae todo de una vez
// y deja que AG Grid virtualice del lado del cliente, acotado por
// `CargaCompleta.truncado` (ver `api/comun.ts`) — a diferencia de
// Historial, esta pantalla no tiene selector de rango de fechas. Auditoría
// genérica (contratistas, empresas, usuarios) desde 2026-08-28 — antes era
// sólo de contratistas.

export type EntidadAuditada = "Contratista" | "Empresa" | "Usuario";

export interface CambioAuditado {
  id: number;
  /** ISO 8601 (UTC) — convertir con `new Date(...)` antes de mostrar. */
  fecha_hora: string;
  usuario_id: number;
  usuario_nombre: string;
  entidad: EntidadAuditada;
  entidad_id: number;
  entidad_nombre: string;
  /** Clave cruda de columna (`"cedula"`, `"nombre"`, `"empresa_id"`,
   * `"tipo_ingreso"`, `"fecha_vencimiento_praind"`, `"es_personal_ruta"`,
   * `"tiene_acceso"`, `"rol"`, `"activo"`, `"password"`) — ver
   * `etiquetaCampo`. */
  campo: string;
  valor_anterior: string | null;
  valor_nuevo: string | null;
}

export function etiquetaEntidad(entidad: EntidadAuditada): string {
  switch (entidad) {
    case "Contratista":
      return "Contratista";
    case "Empresa":
      return "Empresa";
    case "Usuario":
      return "Usuario";
  }
}

/** Espejo de `descripcion_cambio`/`valor_presentable`
 * (`src/comandos/render/auditoria.rs`) — mismas etiquetas, para que la GUI
 * no invente una traducción distinta de las mismas claves crudas. */
export function etiquetaCampo(campo: string): string {
  switch (campo) {
    case "cedula":
      return "Cédula";
    case "nombre":
      return "Nombre";
    case "empresa_id":
      return "Empresa";
    case "tipo_ingreso":
      return "Tipo de ingreso";
    case "fecha_vencimiento_praind":
      return "Vencimiento PRAIND";
    case "es_personal_ruta":
      return "Personal de ruta";
    case "tiene_acceso":
      return "Acceso";
    case "rol":
      return "Rol";
    case "activo":
      return "Activo";
    case "password":
      return "Contraseña";
    default:
      return campo;
  }
}

/** `password` es un marcador de evento sin valores (ver
 * `UsuarioService::cambiar_password_con_hash_auditado` en el núcleo) — sólo
 * importa que ocurrió y cuándo, no hay antes/después que mostrar. */
export function valorPresentable(campo: string, valor: string | null): string {
  if (campo === "password") return "—";
  if (valor === null) {
    return campo === "fecha_vencimiento_praind" ? "Sin fecha" : "—";
  }
  if (campo === "tipo_ingreso" && valor === "IN_HOUSE") return "IN HOUSE";
  if (campo === "tipo_ingreso" && valor === "POR_CORREO") return "POR CORREO";
  if (campo === "fecha_vencimiento_praind") {
    // "YYYY-MM-DD" → "DD/MM/YYYY", como texto — sin pasar por `Date` (mismo
    // motivo que el resto de la app: evita corrimientos de huso horario).
    const partes = valor.split("-");
    if (partes.length === 3) {
      const [anio, mes, dia] = partes;
      return `${dia}/${mes}/${anio}`;
    }
    return valor;
  }
  if (campo === "tiene_acceso" && valor === "HABILITADO") return "Habilitado";
  if (campo === "tiene_acceso" && valor === "DESHABILITADO") return "Deshabilitado";
  if ((campo === "es_personal_ruta" || campo === "activo") && valor === "SI") return "Sí";
  if ((campo === "es_personal_ruta" || campo === "activo") && valor === "NO") return "No";
  return valor;
}

export function listarAuditoria(): Promise<CargaCompleta<CambioAuditado>> {
  return invoke("listar_auditoria");
}

/** Incidentes de gafetes (marcar perdido/resolver) — misma pantalla de
 * Auditoría, tabla de origen distinta (`gafetes_incidentes`, no
 * `auditoria_cambios`), mismo gate (`Operacion::VerAuditoria`). */
export function listarAuditoriaGafetes(): Promise<IncidenteGafete[]> {
  return invoke("listar_auditoria_gafetes");
}

File: src/cli/render/auditoria.rs
//! Tabla de `/auditoria` — cambios auditados de contratistas, empresas y
//! usuarios, sólo lectura. Mismo estilo visual que `/activos` (encabezado en
//! negrita, marcador `›`, pie con el total), pero con columnas fijas: a
//! diferencia de Historial no hay filtro ni exportación, así que no se
//! justifica un selector de columnas (F4) — son siempre las mismas cuatro.

use ratatui::style::{Modifier, Style};
use ratatui::text::{Line, Span};

use crate::cli::columnas::Columna;
use crate::database::queries::auditoria::CambioAuditado;
use crate::tiempo::a_costa_rica;

use super::estilos::{estilo_seleccion, muted};
use super::tabla::{anchos_columnas, fila_columnas};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum ColumnaAuditoria {
    Fecha,
    Entidad,
    Cambio,
    Usuario,
}

impl Columna for ColumnaAuditoria {
    const TODAS: &'static [Self] = &[Self::Fecha, Self::Entidad, Self::Cambio, Self::Usuario];

    fn etiqueta(self) -> &'static str {
        match self {
            Self::Fecha => "Fecha y hora",
            Self::Entidad => "Entidad",
            Self::Cambio => "Cambio realizado",
            Self::Usuario => "Modificado por",
        }
    }

    fn clave(self) -> &'static str {
        match self {
            Self::Fecha => "fecha",
            Self::Entidad => "entidad",
            Self::Cambio => "cambio",
            Self::Usuario => "usuario",
        }
    }
}

fn ancho_fijo(columna: ColumnaAuditoria) -> Option<usize> {
    match columna {
        ColumnaAuditoria::Fecha => Some(18),
        ColumnaAuditoria::Entidad | ColumnaAuditoria::Cambio | ColumnaAuditoria::Usuario => None,
    }
}

fn ancho_maximo(columna: ColumnaAuditoria) -> usize {
    match columna {
        ColumnaAuditoria::Entidad | ColumnaAuditoria::Usuario => 26,
        ColumnaAuditoria::Cambio => 60,
        ColumnaAuditoria::Fecha => 18,
    }
}

fn valor(item: &CambioAuditado, columna: ColumnaAuditoria) -> String {
    match columna {
        ColumnaAuditoria::Fecha => a_costa_rica(item.fecha_hora)
            .format("%d/%m/%Y %H:%M")
            .to_string(),
        ColumnaAuditoria::Entidad => format!("{} ({})", item.entidad_nombre, texto_entidad(item)),
        ColumnaAuditoria::Cambio => descripcion_cambio(item),
        ColumnaAuditoria::Usuario => item.usuario_nombre.clone(),
    }
}

fn texto_entidad(item: &CambioAuditado) -> &'static str {
    use crate::database::queries::auditoria::EntidadAuditada;
    match item.entidad {
        EntidadAuditada::Contratista => "contratista",
        EntidadAuditada::Empresa => "empresa",
        EntidadAuditada::Usuario => "usuario",
    }
}

/// Reescrito de `tui::auditoria::render::descripcion_cambio` (DEC-002/
/// DEC-014: `src/tui/` no se toca ni se comparte) — mismas etiquetas y
/// formato "Campo: antes → después", salvo `password` (marcador de evento
/// sin valores, ver `UsuarioService::cambiar_password_con_hash_auditado`):
/// ahí no hay antes/después que mostrar, sólo que ocurrió.
fn descripcion_cambio(item: &CambioAuditado) -> String {
    if item.campo == "password" {
        return "Contraseña actualizada".to_owned();
    }
    let etiqueta = match item.campo.as_str() {
        "cedula" => "Cédula",
        "nombre" => "Nombre",
        "empresa_id" => "Empresa",
        "tipo_ingreso" => "Tipo de ingreso",
        "fecha_vencimiento_praind" => "Vencimiento PRAIND",
        "es_personal_ruta" => "Personal de ruta",
        "tiene_acceso" => "Acceso",
        "rol" => "Rol",
        "activo" => "Activo",
        _ => item.campo.as_str(),
    };
    let anterior = valor_presentable(&item.campo, item.valor_anterior.as_deref());
    let nuevo = valor_presentable(&item.campo, item.valor_nuevo.as_deref());
    format!("{etiqueta}: {anterior} → {nuevo}")
}

fn valor_presentable(campo: &str, valor: Option<&str>) -> String {
    let Some(valor) = valor else {
        return if campo == "fecha_vencimiento_praind" {
            "Sin fecha".to_owned()
        } else {
            "—".to_owned()
        };
    };
    match (campo, valor) {
        ("tipo_ingreso", "IN_HOUSE") => "IN HOUSE".to_owned(),
        ("tipo_ingreso", "POR_CORREO") => "POR CORREO".to_owned(),
        ("fecha_vencimiento_praind", valor) => chrono::NaiveDate::parse_from_str(valor, "%Y-%m-%d")
            .map_or_else(
                |_| valor.to_owned(),
                |fecha| fecha.format("%d/%m/%Y").to_string(),
            ),
        ("tiene_acceso", "HABILITADO") => "Habilitado".to_owned(),
        ("tiene_acceso", "DESHABILITADO") => "Deshabilitado".to_owned(),
        ("es_personal_ruta" | "activo", "SI") => "Sí".to_owned(),
        ("es_personal_ruta" | "activo", "NO") => "No".to_owned(),
        _ => valor.to_owned(),
    }
}

pub(super) fn lineas_tabla_auditoria(
    items: &[CambioAuditado],
    total: usize,
    seleccion: usize,
    ancho: u16,
) -> Vec<Line<'static>> {
    let anchos = anchos_columnas(
        ancho,
        ColumnaAuditoria::TODAS.iter().copied(),
        ancho_fijo,
        ancho_maximo,
    );
    let mut lineas = vec![
        Line::from(Span::styled(
            format!(
                "  {}",
                fila_columnas(&anchos, |_| false, |c| c.etiqueta().to_uppercase())
            ),
            Style::default().add_modifier(Modifier::BOLD),
        )),
        Line::from(Span::styled("─".repeat(ancho as usize), muted())),
    ];
    for (indice, item) in items.iter().enumerate() {
        let marcador = if indice == seleccion { "› " } else { "  " };
        let texto = format!(
            "{marcador}{}",
            fila_columnas(&anchos, |_| false, |c| valor(item, c))
        );
        lineas.push(if indice == seleccion {
            Line::from(Span::styled(texto, estilo_seleccion()))
        } else {
            Line::from(texto)
        });
    }
    if items.is_empty() {
        lineas.push(Line::from(Span::styled("Sin cambios auditados", muted())));
    }
    lineas.push(Line::from(""));
    lineas.push(Line::from(Span::styled(cantidad_cambios(total), muted())));
    lineas
}

fn cantidad_cambios(total: usize) -> String {
    if total == 1 {
        "1 cambio auditado".to_string()
    } else {
        format!("{total} cambios auditados")
    }
}
